Radio station

bryan-hub opened the pull request:
# Description

Added files and edits to create a working radio station. Includes new channel, PBS, and one way tabletop radios.

Partmedia commented on Content.Server/Radio/EntitySystems/RadioDeviceSystem.cs:
```diff
@@ -216,9 +216,17 @@ private void OnReceiveRadio(EntityUid uid, RadioSpeakerComponent component, ref
+        }
+    }
+
+    private string ToString(RadioChannelPrototype channel)
+    {
+        throw new NotImplementedException();
```
This is no longer needed, please remove this.

bryan-hub replied:
I removed the unneeded code block bracket and moved the indent left one tab

Partmedia replied:
The unneeded `ToString()` is still here.

bryan-hub replied:
I didn't add that

Partmedia replied:
It still shows up in the diff -- and will be added if this is merged. If it's not being used, please remove it.

bryan-hub replied:
gotcha, okay I removed it

Partmedia commented on Content.Server/Radio/EntitySystems/RadioDeviceSystem.cs:
```diff
@@ -216,9 +216,17 @@ private void OnReceiveRadio(EntityUid uid, RadioSpeakerComponent component, ref
-        // log to chat so people can identity the speaker/source, but avoid clogging ghost chat if there are many radios
-        var message = args.OriginalChatMsg.Message; // The chat system will handle the rest and re-obfuscate if needed.
-        _chat.TrySendInGameICMessage(uid, message, InGameICChatType.Whisper, ChatTransmitRange.GhostRangeLimit, nameOverride: name, checkRadioPrefix: false, languageOverride: args.Language);
+        {
```
Why is this in a new scoped block?

bryan-hub replied:
I had set the files aside when I moved to the new PR branch, must've copied it wrong, I'll push the proper code

Partmedia commented on Content.Server/Radio/Components/RadioSpeakerComponent.cs:
```diff
@@ -24,4 +24,8 @@ public sealed partial class RadioSpeakerComponent : Component
     [DataField("enabled")]
     public bool Enabled;
+
+    [DataField("IsSpeaker")]
```
Please add a comment explaining the purpose of this new data field.

bryan-hub replied:
comment added

Partmedia reviewed (changes requested):
Thanks for the PR. Please see review comments inline.

Partmedia commented on Content.Server/Radio/Components/RadioSpeakerComponent.cs:
```diff
@@ -24,4 +24,8 @@ public sealed partial class RadioSpeakerComponent : Component
     [DataField("enabled")]
     public bool Enabled;
+
+    [DataField("IsSpeaker")] //This field when true will play the audio over the Speak chat type. The default chat type is whisper. This will allow the audio to carry further.
```
```suggestion
    [DataField("IsSpeaker")] // Set to true for broadcast radio speakers so that it speaks instead of whispers the message.
```

Partmedia commented on Resources/Prototypes/_Nuclear14/Entities/Structures/Misc/radio.yml:
```diff
@@ -18,4 +18,140 @@
\ No newline at end of file
+        - LowImpassable
+
+- type: entity #prototype for radio station microphone
```
```suggestion
- type: entity # radio station microphone
```

Partmedia commented on Resources/Prototypes/_Nuclear14/Entities/Structures/Misc/radio.yml:
```diff
@@ -18,4 +18,140 @@
+    state: radiomicrophone
+  - type: RadioMicrophone
+    broadcastChannel: PBS
+    toggleOnInteract: false # Nuclear-14
+    frequency: 1667 # Nuclear-14
+    listenRange: 2
+  - type: Clickable
+  - type: InteractionOutline
+  - type: Appearance
+  - type: WiresVisuals
+  - type: Transform
+    noRot: false
+    anchored: true
+  - type: ActiveRadio
+    channels:
+    - PBS
+  - type: Fixtures
+    fixtures:
+      fix1:
+        shape:
+          !type:PhysShapeCircle
+            radius: 0.20
+        density: 1000
+        mask:
+        - MachineMask
+        layer:
+        - MachineLayer
+  # Nuclear-14-Start
+  - type: ActivatableUI
+    key: enum.HandheldRadioUiKey.Key
+  - type: UserInterface
+    interfaces:
+      - key: enum.HandheldRadioUiKey.Key
+        type: HandheldRadioBoundUserInterface
+  # Nuclear-14-End
```
This whole section is for N14, so you don't need the `# Nuclear-14` comments

bryan-hub replied:
those were artifacts from copy/pasting the code, I'll remove and push

Partmedia commented on Resources/Prototypes/_Nuclear14/Entities/Structures/Misc/radio.yml:
```diff
@@ -18,4 +18,140 @@
+    channels:
+    - PBS
+    enabled: true
+    toggleOnInteract: true # Nuclear-14
+    IsSpeaker: true
+  - type: Tag
+    tags:
+    - Radio
+  - type: Speech
+    speechVerb: Robotic
+  - type: ActiveRadio
+    channels:
+    - PBS
+    # Nuclear-14-Start
+  - type: RadioMicrophone
+    broadcastChannel: PBS
+    toggleOnInteract: false # Nuclear-14
+    frequency: 1667 # Nuclear-14
```
Same as above

Partmedia commented on Resources/Prototypes/_Nuclear14/Entities/Structures/Misc/radio.yml:
```diff
@@ -18,4 +18,140 @@
+    state: radiomicrophonestand
+  - type: RadioMicrophone
+    broadcastChannel: PBS
+    toggleOnInteract: false # Nuclear-14
+    frequency: 1667 # Nuclear-14
+    listenRange: 2
+  - type: ActiveRadio
+    channels:
+    - PBS
+    listenRange: 2
+  - type: Fixtures
+    fixtures:
+      fix1:
+        shape:
+          !type:PhysShapeAabb
+          bounds: "-0.25,0.5,0,-0.5"
+        layer:
+        - LowImpassable
+  # Nuclear-14-Start
+  - type: ActivatableUI
+    key: enum.HandheldRadioUiKey.Key
+  - type: UserInterface
+    interfaces:
+      - key: enum.HandheldRadioUiKey.Key
+        type: HandheldRadioBoundUserInterface
+  # Nuclear-14-End
```
Same as above

bryan-hub commented:
Is there anything else needed before merge?

Partmedia commented:
A few more comments:
- The standing microphone does not default to the broadcast frequency. Not sure if this was intended or not.
- An exception shows up on the server when you attempt to turn on the (non-existent) speaker button on the radio UI:
  ```
  [ERRO] system.radio_device: Can't resolve "Content.Server.Radio.Components.RadioSpeakerComponent" on entity radio microphone (43074/n43074, RadioDJMicrophone)!
   at Content.Server.Radio.EntitySystems.RadioDeviceSystem.SetSpeakerEnabled(EntityUid uid, EntityUid user, Boolean enabled, Boolean quiet, RadioSpeakerComponent component) in Content.Server/Radio/EntitySystems/RadioDeviceSystem.cs:line 158
   ```
- An FTL string is missing for the PBS channel `chat-radio-pbs` (see screenshot)
- The GUI title still shows up as "Handheld radio" (see screenshot)

![image](https://github.com/user-attachments/assets/e53b06a4-2b9a-4e71-812f-2fe615f25699)

Other than that, this seems to work.

bryan-hub commented:
Added the 1667 frequency to the standing mics, tested. Removed UI from mic's all together. It is unnecessary since you no longer need to change the frequency or speaker.  This way, if you click the mic, it enables it. This removes the error and the UI name, tested as well. Also, added pbs as Public Broadcast. Should hopefully address everything. Tested.
